File: src/main/java/dev/stocky37/epic7/core/Stat.java
package dev.stocky37.epic7.core;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;


public enum Stat {
	ATK("atk", "Attack", false, true),
	ATK_PERC("atk%", "Attack%", true, true, Stat.ATK),
	HP("hp", "Health", false, true),
	HP_PERC("hp%", "Health%", true, true, Stat.HP),
	DEF("def", "Defense", false, true),
	DEF_PERC("def%", "Defense%", true, true, Stat.DEF),
	SPD("spd", "Speed", false, true),
	SPD_PERC("spd%", "Speed%", true, true, Stat.SPD),
	CHC("chc", "Critical Hit Chance", true, true),
	CHD("chd", "Critical Hit Damage", true, true),
	EFF("eff", "Effectiveness", true, true),
	EFR("efr", "Effect Resistance", true, true),
	DAC("dac", "Dual Attack Chance", true, false),
	LS("ls", "Lifesteal", true, false),
	CA("ca", "Counterattack", true, false),
	DMG("dmg", "Damage", true, false);


	private final String id;
	private final String name;
	private final boolean percentage;
	private final boolean availableOnGear;
	private final Stat baseStat;

	Stat(String id, String name, boolean percentage, boolean availableOnGear) {
		this(id, name, percentage, availableOnGear, null);
	}

	Stat(String id, String name, boolean percentage, boolean availableOnGear, Stat baseStat) {
		this.id = id;
		this.name = name;
		this.percentage = percentage;
		this.availableOnGear = availableOnGear;
		this.baseStat = baseStat;
	}

	public static List<Stat> getAllBaseStats() {
		return Arrays.stream(Stat.values())
				.filter(Stat::isBaseStat)
				.collect(Collectors.toList());
	}

	public static List<Stat> getAllGearStats() {
		return Arrays.stream(Stat.values())
				.filter(Stat::isAvailableOnGear)
				.collect(Collectors.toList());
	}

	// todo: if this is slow, change to map impl
	public static Stat fromId(String str) throws IllegalArgumentException {
		return Arrays.stream(Stat.values())
				.filter(stat -> stat.getId().equalsIgnoreCase(str))
				.findAny()
				.orElseThrow(IllegalArgumentException::new);
	}

	public String getId() {
		return id;
	}

	public String getName() {
		return name;
	}

	public boolean isPercentage() {
		return percentage;
	}

	public boolean isAvailableOnGear() {
		return availableOnGear;
	}

	public Stat getBaseStat() {
		return baseStat == null ? this : baseStat;
	}

	public boolean isBaseStat() {
		return baseStat == null;
	}
}

File: src/main/kotlin/dev/stocky37/epic7/core/Stat.kt
package dev.stocky37.epic7.core

enum class Stat constructor(
	val id: String,
	val label: String,
	val isPercentage: Boolean,
	val isAvailableOnGear: Boolean,
	private val baseStat: Stat? = null
) {
	ATK("atk", "Attack", false, true),
	ATK_PERC("atk%", "Attack%", true, true, Stat.ATK),
	HP("hp", "Health", false, true),
	HP_PERC("hp%", "Health%", true, true, Stat.HP),
	DEF("def", "Defense", false, true),
	DEF_PERC("def%", "Defense%", true, true, Stat.DEF),
	SPD("spd", "Speed", false, true),
	SPD_PERC("spd%", "Speed%", true, true, Stat.SPD),
	CHC("chc", "Critical Hit Chance", true, true),
	CHD("chd", "Critical Hit Damage", true, true),
	EFF("eff", "Effectiveness", true, true),
	EFR("efr", "Effect Resistance", true, true),
	DAC("dac", "Dual Attack Chance", true, false),
	LS("ls", "Lifesteal", true, false),
	CA("ca", "Counterattack", true, false),
	DMG("dmg", "Damage", true, false);

	val isBaseStat: Boolean by lazy {
		baseStat == null
	}

	fun getBaseStat(): Stat {
		return baseStat ?: this
	}

	companion object {
		fun getBaseStats(): List<Stat> {
			return Stat.values().filter { it.isBaseStat }
		}

		fun getGearStats(): List<Stat> {
			return Stat.values().filter { it.isAvailableOnGear }
		}

		// todo: if this is slow, change to map impl
		fun fromId(str: String): Stat {
			return values().single { stat -> stat.id.equals(str, ignoreCase = true) }
		}
	}
}



